src/webgl/p5.Matrix.js

```diff
@@ -1,717 +1,1240 @@
+ *
+ *   new p5.Matrix(p5.Matrix | Float32Array | Array); // copy
+ *
+ *   new p5.Matrix(p5); // new 4x4 matrix + p5-reference
+ *
+ * </pre>
+ *
  * @class p5.Matrix
  * @private
  * @constructor
- * @param {Array} [mat4] array literal of our 4x4 matrix
+ * @param {...*} variable args
+ * @chainable
  */
 p5.Matrix = function() {
   var args = new Array(arguments.length);
   for (var i = 0; i < args.length; ++i) {
     args[i] = arguments[i];
   }
 
-  // This is default behavior when object
-  // instantiated using createMatrix()
-  // @todo implement createMatrix() in core/math.js
+  // check if last arg is p5-reference
   if (args.length && args[args.length - 1] instanceof p5) {
     this.p5 = args[args.length - 1];
   }
 
-  if (args[0] === 'mat3') {
-    this.mat3 = Array.isArray(args[1])
-      ? args[1]
-      : new GLMAT_ARRAY_TYPE([1, 0, 0, 0, 1, 0, 0, 0, 1]);
+  // assume, args[0] is a (typed) array[16]
+  var mat;
+
+  // 1) copy constructor
+  if (args[0] instanceof p5.Matrix) {
+    var other = args[0];
+    if (other.p5) this.p5 = other.p5;
+    if (other.mat3) this.mat3 = other.mat3.slice();
+    if (other.mat4) this.mat4 = other.mat4.slice();
+  } else if (args[0] === 'mat3') {
+    // 2) 3x3 matrix: check for 'mat3' and eventually given array
+    // assume, args[1] is a (typed) array[9]
+    mat = args[1];
+
+    // if no valid array, create 3x3 identity
+    // note: .length check, instead of .constructor check.
+    if (!mat || mat.length !== 9) {
+      mat = [1, 0, 0, 0, 1, 0, 0, 0, 1];
+    }
+    // if available create as Float32Array
+    if (Float32Array) {
+      mat = new Float32Array(mat);
+    }
+    this.mat3 = mat;
   } else {
-    this.mat4 = Array.isArray(args[0])
-      ? args[0]
-      : new GLMAT_ARRAY_TYPE([1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1]);
+    // 3) 4x4 matrix: this is basically default when no args are given
+    // assume, args[0] is a (typed) array[16]
+    mat = args[0];
+
+    // if no valid array, create 4x4 identity
+    // note: .length check, instead of .constructor check.
+    if (!mat || mat.length !== 16) {
+      mat = [1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1];
+    }
+    // if available create as Float32Array
+    if (Float32Array) {
+      mat = new Float32Array(mat);
+    }
+    this.mat4 = mat;
   }
 
   this.name = 'p5.Matrix'; // for friendly debugger system
   return this;
 };
 
 /**
- * Sets the x, y, and z component of the vector using two or three separate
- * variables, the data from a p5.Matrix, or the values from a float array.
+ * Creates and returns a new (deep) copy of this matrix.
+ * A copy can also be made by using the copy-constructor.
  *
- * @method set
- * @param {p5.Matrix|Float32Array|Number[]} [inMatrix] the input p5.Matrix or
- *                                     an Array of length 16
+ * @method copy
+ * @return {p5.Matrix} matrix
  * @chainable
  */
+p5.Matrix.prototype.copy = function() {
+  return new p5.Matrix(this);
```

Comment:
you're losing the p5 instance here.

Reply:
The p5.Matrix is copied in the constructor, this includes the p5 reference, if there was one.

However, this is not very well designed anyways. 
A Matrix should do matrix stuff and not care about any deg-to-rad-etc-conversion - which is the only reason for the p5-ref in the first place. 
IMO, in a next step, the p5 member should be completely removed from the matrix or any other specialized kind for that matter.
Apart from that, it is valid to create a matrix without any p5-reference and then apply/set it as one of the core-matrices (modelview, projection, etc...). In such a case the reference is lost by all means. As a result, code and execution and will be just a mess.

The deg/rad-mode option (if useful at all) should be handled by a layer (p5 itself) that is responsible for that option.

Reply:
Oh yes sorry about that.

Another thing: can you specify the parameter arguments fully (i.e. No varargs). You can add an overload if there's an ambiguity.

Reply:
Updated. But I'm not sure whats the preferred way of documenting such overloaded functions.